src/azul/terra.py

```diff
@@ -195,10 +197,25 @@ class TDRClient(SAMClient):
     A client for the Broad Institute's Terra Data Repository aka "Jade".
     """
 
+    def project_for_source(self, source: TDRSource) -> str:
+        """
+        Use the TDR API to find which Google Cloud project contains the source.
+        """
+        response = self._get_source_response(source)
+        assert response.status == 200
+        response = json.loads(response.data)
+        return response['dataProject']
+
     def check_api_access(self, source: TDRSource) -> None:
         """
         Verify that the client is authorized to read from the TDR service API.
         """
+        response = self._get_source_response(source)
+        assert response.status == 200
+        log.info('TDR client is authorized for API access to %s: %r',
+                 source, response.data[:256].decode() + '...')
```

Comment:
You have to decode first then truncate, otherwise you risk cutting right through a character that requires multiple bytes in UTF-8.

Reply:
And use same truncation rules as we used for the BQ logging, except for 256 instead of 2048 as a threshold.